### DISCUSSION TITLE
How can I Limit content in Terminal?

### DISCUSSION THREAD

**Discussion by bajrangCoder** [sourcelair/xterm.js#4558] [Q&A]
I have created a terminal using xtermjs and node-pty, which works well but When terminal have 100 + columns then it's lag so i want to limit the column and if it exceeds the limit then just start remove column from starting.
For eg:
Assume these numbers as column:(and max column is 5)
1
2
3
4
5
And now adds a new data in terminal 6 and remove 1 then it looks like : 2, 3, 4, 5, 6

How can I do this ? I have tried some ways but failed. please help me? VScode also do this , I thought it uses scrolllines(-ve value);
?

**Comment by jerch** [comment]
Did you mix up "columns" and "rows" in your question?

column output can only be changed via resize (just make the terminal smaller in width), it is not possible to truncate column output by other means.
Visible rows can also be changed by resize. Additionally the terminal maintains a scrollback setting to limit max rows held by the terminal. Normally you dont need to change the scrollback for perf reasons, it has lit. no impact on perf.

A terminal (with or w'o scrollback) already implements the "windowing" on rows as you describe with your numbering scheme.

If output lags alot, maybe try a different renderer (canvas or webgl addon). If output still lags with one of these, then the issue is most likely within your code.

**Reply by bajrangCoder** [reply]
By column, I mean lines . 
I'm using webgl addon.
Let me explain what I wanted to do:
In terminal if there is tone of content (lines in terminal) it affect the performance. Many terminals also remove old or previous lines from terminal if it exceeds it's max line limit in terminal (for eg: Termux, Arch Linux terminal , etc) .
I also want to implement the same in my Terminal, By changing rows /cols it doesn't affect. Is there any good approach to do this in xtermjs

**Reply by Tyriar** [reply]
@bajrangCoder you want your rows and columns to be the size of the grid of characters that the user will see (ie. the viewport). The scrollback is how many lines we keep track when they scroll off the viewport. If you run `ls` on https://xtermjs.org/, you will see some lines scroll off, they're going to the scrollback.

> I'm using webgl addon.

If you make rows and columns too large, you could run into problems with the browser's GPU support, such as maximum texture size.

**Reply by bajrangCoder** [reply]
Ok Got it.
Thanks ❤️
One more question:
Is there any way to give Autocompletion in xtermjs terminal, same as fish shell .
If it's possible then provide some hint .

**Reply by jerch** [reply]
A terminal is a terminal, and a shell is a shell, they are not the same. You can run fish from xterm.js, so this would give you fish-like auto-completion. A terminal on its own cannot do that. So the answer is yes and no.

**Reply by bajrangCoder** [reply]
Noo, I don't want to make totally like fish.
Just tell Is it possible to show Autocompletion in xtermjs means Autocompletion are in dim color.
For eg:
~ $ cd {and this is suggested my terminal in dim color}

**Reply by jerch** [reply]
Not sure what you are looking for, let me try to answer it differently - if your favorite desktop terminal emulator can do it, most likely xterm.js can do it as well. (Unless its a cutting edge feature only supported by your desktop terminal).

**Reply by bajrangCoder** [reply]
Ok 😓
